src/sidebar/test/service-url-test.js

```diff
@@ -1,20 +1,166 @@
 'use strict';
 
-var serviceUrl = require('../service-url');
+var proxyquire = require('proxyquire');
 
-describe('serviceUrl', function () {
-  var service;
+/** Return a fake annotationUI object. */
+function fakeAnnotationUI() {
+  var links = null;
+  return {
+    updateLinks: function(newLinks) {
+      links = newLinks;
+    },
+    getState: function() {
+      return {links: links};
+    },
+  };
+}
 
-  beforeEach(function () {
-    service = serviceUrl({serviceUrl: 'https://test.hypothes.is/'});
+function createServiceUrl(linksPromise) {
+  var replaceURLParams = sinon.stub().returns(
+    {url: 'EXPANDED_URL', params: {}}
+  );
+
+  var serviceUrlFactory = proxyquire('../service-url', {
+    './util/url-util': { replaceURLParams: replaceURLParams },
+  });
+
+  var annotationUI = fakeAnnotationUI();
+
+  var store = {
+    links: sinon.stub().returns(linksPromise),
+  };
+
+  return {
+    annotationUI: annotationUI,
+    store: store,
+    serviceUrl: serviceUrlFactory(annotationUI, store),
+    replaceURLParams: replaceURLParams,
+  };
+}
+
+describe('links', function () {
+
+  beforeEach(function() {
+    sinon.stub(console, 'warn');
+  });
+
+  afterEach(function () {
+    console.warn.restore();
+  });
+
+  context('before the API response has been received', function() {
+    var serviceUrl;
+    var store;
+
+    beforeEach(function() {
+      // Create a serviceUrl function with an unresolved Promise that will
+      // never be resolved - it never receives the links from store.links().
+      var parts = createServiceUrl(new Promise(function() {}));
+
+      serviceUrl = parts.serviceUrl;
+      store = parts.store;
+    });
+
+    it('sends one API request for the links at boot time', function() {
+      assert.calledOnce(store.links);
+      assert.isTrue(store.links.calledWithExactly());
+    });
+
+    it('returns an empty string for any link', function() {
+      assert.equal(serviceUrl('foo'), '');
+    });
+
+    it('returns an empty string even if link params are given', function() {
+      assert.equal(serviceUrl('foo', {bar: 'bar'}), '');
+    });
   });
 
-  it('returns route URLs', function () {
-    assert.equal(service('help'), 'https://test.hypothes.is/docs/help');
+  context('if the API request fails', function() {
+    it('just keeps returning empty strings for URLs', function() {
+      var linksPromise = Promise.reject(new Error('Oops'));
+
+      var serviceUrl = createServiceUrl(linksPromise).serviceUrl;
+
+      assert.equal(serviceUrl('second_link'), '');
```

Comment:
This assert is being executed _before_ the Promise handlers inside `serviceUrl` have actually been executed, so this isn't testing the behaviour you intended to test. The behaviour you actually want to test is difficult without the kind of refactoring we talked about in Slack though to make it possible to wait on the link fetch completing.